### DISCUSSION TITLE
Leaderless QQs reporting noproc

### DISCUSSION THREAD

**Discussion by Ayanda-D** [rabbitmq/rabbitmq-server#12701] [Other] [answered]
### Community Support Policy

- [X] I have read [RabbitMQ's Community Support Policy](https://github.com/rabbitmq/rabbitmq-server/blob/main/COMMUNITY_SUPPORT.md)

### RabbitMQ version used

3.13.6

### How is RabbitMQ deployed?

Debian package

### Steps to reproduce the behavior in question

Hi Team 👋 

We've been running a number of test scenarios over the last couple of months to evaluate QQ performance and resiliency (before upgrading some of our client critical CMQs) and sometimes (not often) we encounter leaderless QQs, where leaders become unavailable and are reported as `noproc` despite sanity checks such as `check_if_node_is_quorum_critical` commands succeeding. On our maintenance and upgrade procedures, we consider it a big risk to restart brokers and machines on which leaders are expected to be running but wouldn't, which would result in message loss. Is this a known issue? This looks very similar QQ leader behaviour described for followers here https://github.com/rabbitmq/rabbitmq-server/discussions/10423 (although we are prioritizing addressing the leaders, from which we'd still access the queue).

We have been looking at and experimenting with a custom workaround CLI command which (efficiently) checks sanity of leaders, which we're looking to PR for comments and review, and to have available for use in later rabbit releases. Plan is to optionally use this command along with the already provided upgrade and maintenance queue checks when upgrading critical environments. Command signature we have so far (open to other name suggestions 🙏 ):

```
rabbitmq-queues leader_health_check [--vhost <vhost>] [--global] <pattern>
```

which spawns concurrent leader checks (with concurrency limit/threshold which we dont want to exceed), where:

```
["<pattern>", "regular expression pattern used to match quorum queues"],
["--global", "run leader health check for all queues in all virtual hosts on the node"],
```

We'll follow up with the implementation for further comments. This is to mitigate the major risk factor of `noproc` leaders (of potential message loss), which is currently a blocker on our CMQs -> QQs upgrade path.

**Comment by kjnilsson** [comment]
I think the first thing to do is to provide information on how these queues failed so that we can fix that (if still an issue).

**Reply by kjnilsson** [reply]
did you use exit(Pid, kill) to simulate process crashes?

**Reply by Ayanda-D** [reply]
We see this on very rare occasions, unfortunately haven't been able to consistently reproduce. Didn't manage to simulate/reproduce this with `exit(Pid, kill)` either (ra_state/leaderboards seem to be updated as expected). Bit difficult to see the root cause, but having seen it, couldnt shrug it off & looked at the workaround cli command to protect things a bit on restarts.

**Reply by kjnilsson** [reply]
I suspect that this could be improved by using `whereis(K)` [here](https://github.com/rabbitmq/rabbitmq-server/blob/8dd49a532d2ec05b5543efc518b2fc9f337c580b/deps/rabbit/src/rabbit_quorum_queue.erl#L449) and updating the status accordingly. that is what the quorum_status command does.

**Reply by Ayanda-D** [reply]
Yep, I think that will also improve the issue, I've just tried it out. I'll forward a PR for this as well soon. The only other thing that the cli command does is an explicit `ra_server_proc:ping/1` per leader (slightly more expensive). Let me prepare these and forward for review 👍

**Reply by kjnilsson** [reply]
cool. ping/1 will be much too expensive for large numbers of QQs and this should provide much of the same guarantees whilst being pretty cheap.

**Reply by Ayanda-D** [reply]
Ok cool. i've opened https://github.com/rabbitmq/rabbitmq-server/pull/12709 - please check. i'm ignoring [this clause](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_quorum_queue.erl#L452-L453) - i'm not seeing where/how `ra_state` can hold anything else? Unless this was for backward compatibility (or should we let it crash if [anything else](https://github.com/rabbitmq/rabbitmq-server/pull/12709/files#diff-aafd1be4bf1d8c8af7462438372594b2b2253cfb3187687df6d30f92526e2f75R452)).

**Comment by michaelklishin** [comment]
Keep in mind that RabbitMQ 3.13.x is [out of community support](https://www.rabbitmq.com/release-information), so any contributions in this area won't be available in OSS RabbitMQ 3.13.x.

**Reply by Ayanda-D** [reply]
Ok 👍

**Comment by michaelklishin** [comment, marked as answer]
This should be addressed by https://github.com/rabbitmq/rabbitmq-server/pull/12727 (contributed by @Ayanda-D) and available for testing in alpha [`4.0.4-alpha.0caa0664`](https://github.com/rabbitmq/server-packages/releases/tag/alphas.1731713441725).
